### ISSUE TITLE
CertificatePinner without pins

### ISSUE THREAD

**Issue by ikoz** [square/okhttp#1947]
If a developer - for whatever reason - forgets to add pins to CertificatePinner, all connections will succeed as if CertificatePinner was not used - as long as a trust anchor is in the system's trust store.

For example:
```java
okhttpClient.setCertificatePinner(CertificatePinner.DEFAULT);
```
or
```java
CertificatePinner certificatePinner = new CertificatePinner.Builder().build();
okhttpClient.setCertificatePinner(certificatePinner);
```

This is probably not really a bug. However this would be a critical issue if discovered on someone's code (and it has). 

I think there should be some warning and/or this behavior should be documented.
One could argue that the CertificatePinner should always fail if there are no pins (fail-open vs fail-close)

This happens because of the following code in CertificatePinner

```java
  public void check(String hostname, List<Certificate> peerCertificates)
      throws SSLPeerUnverifiedException {

    Set<ByteString> pins = findMatchingPins(hostname);

    if (pins == null) return;
```
where findMatchingPins() returns null if hostnameToPins is empty.

```java
  Set<ByteString> findMatchingPins(String hostname) {
    Set<ByteString> directPins   = hostnameToPins.get(hostname);
    if (directPins == null && wildcardPins == null) return null;
```

**Comment by swankjesse** [comment]
Would it help if you could add a pin for `*` ? You could potentially pin it to something that won't match any certificate.

**Comment by ikoz** [comment]
I assume that if a pin is added for` *` that does not match anything, then the implementation will fail all connections.

However this is not the issue here.

The issue I am highlighting is that some might forget to add any pins (as I describe, he could just use the .DEFAULT instance).

Without pins, all connections will work normally - nothing will fail. A developer will end up thinking he is using pinning, while in reality he isn't.

CertificatePinner should *fail* to establish connections if there are no pins set - instead of allowing all to succeed.

**Comment by swankjesse** [comment]
If you forget to add pins, you're still protected by CA certificate. I don't think there's an attack vector here. Furthermore, I definitely do not want to encourage people to pin certificates hastily. Adding a pin is extremely dangerous because it risks severe availability problems.

**Comment by ikoz** [comment]
Indeed, if you forget to add pins, as long as there's a trust anchor in the trust store the connection is secure; that was never questioned. Never mentioned there's an attack vector or a bug here.

The issue here is behaving in an expected way. 
In my opinion, if someone purposely does a setCertificatePinner() call, he really wants Certificate Pinning to work; He wants connections to *always fail* by default unless explicitly allowed using by pin verification.
To that end, an empty pinset should mean that all connections fail - not that all succeed as currently happens.
Anyway, it's really a matter of semantics and attitude towards default-fail vs default-allow. Essentially, helping developers avoid errors instead of encouraging errors.
Thanks.

**Comment by artem-zinnatullin** [comment]
So, if I want to use same OkHttpClient instance for requests to different
hosts and pin certificate(s) only to one/several hosts how should I do it
with suggested behaviour?

On Thu, Oct 29, 2015, 01:48 John Kozyrakis <notifications@github.com> wrote:

> Indeed, if you forget to add pins, as long as there's a trust anchor in
> the trust store the connection is secure; that was never questioned. Never
> mentioned there's an attack vector or a bug here.
>
> The issue here is behaving in an expected way.
> In my opinion, if someone purposely does a setCertificatePinner() call, he
> really wants Certificate Pinning to work; He wants connections to *always
> fail* by default unless explicitly allowed using by pin verification.
> To that end, an empty pinset should mean that all connections fail - not
> that all succeed as currently happens.
> Anyway, it's really a matter of semantics and attitude towards
> default-fail vs default-allow. Essentially, helping developers avoid errors
> instead of encouraging errors.
> Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/okhttp/issues/1947#issuecomment-152020970>.
>
-- 

@artem_zin

**Comment by ikoz** [comment]
That's an interesting use case I had not thought of. 

If that feature is needed, my solution, thinking primarily about security, would be a whitelist approach. 

However, I do see how this can be complicated.

One would have to know which domains are non-pinned, and add them along with the pins, with an empty pin or some specific ALLOW identifier.

For example
```java
     CertificatePinner certificatePinner = new CertificatePinner.Builder()
         .add(hostnameA, "sha1/BOGUSPIN")
         ​.add(hostnameB, "sha1/BOGUSPIN")
         ​.add(hostnameC, "ALLOW")
         .build();
````
In this implementation, hostnameA and hostnameB would only work with that pin, hostnameC will pass-through pinner, all other domains will fail.

**Comment by swankjesse** [comment]
What happens if a request to hostnameC redirects to hostnameD ?

**Comment by ikoz** [comment]
Conceptually 'pinning'  is orthogonal to redirects. 

Each connection should be assessed on its own against the pins, even if it comes from a redirect.

So to answer your question, connections to hostnameD should fail, as you wanted a pinned/increased level of trust from that connection, but you will not get it.

It's typical for pinned connections not to succeed if they redirect to unpinned hosts, across all pinning implementations I've reviewed.

**Comment by ikoz** [comment]
Another option would be to explicitly allow domains without pins through the pinner.

For example:

```java
     CertificatePinner certificatePinner = new CertificatePinner.Builder()
         .add(hostnameA, "sha1/BOGUSPINA")
         ​.add(hostnameB, "sha1/BOGUSPINB")
         ​.allowNonPinnedDomains(True)
         .build();
```

**Comment by artem-zinnatullin** [comment]
Too much combinations of possible requirements for simple
CertificatePinner. What if I'd like to work with some hostname without
pinning (https://google.com, etc) but prevent requests to other non-pinned
hosts. And I see other strange combinations of possible
problems/requirements.

I mean, CertificatePinner has one job — pin certificates to hostnames.

You can use Interceptors API to prevent/allow requests to different hosts
with fully custom logic.

On Thu, Oct 29, 2015, 21:17 John Kozyrakis <notifications@github.com> wrote:

> Another option would be to explicitly allow domains without pins through
> the pinner.
>
> For example:
>
>      CertificatePinner certificatePinner = new CertificatePinner.Builder
>
> ()
>          .add(hostnameA, "sha1/BOGUSPINA")
>          ​.add(hostnameB, "sha1/BOGUSPINB")
>          ​.allowNonPinnedDomains(True)
>          .build();
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/okhttp/issues/1947#issuecomment-152272221>.
>
-- 

@artem_zin

**Comment by ikoz** [comment]
Here you go:

```java
     CertificatePinner certificatePinner = new CertificatePinner.Builder()
         .add(hostnameA, "sha1/BOGUSPINA")
         ​.add(hostnameB, "sha1/BOGUSPINB")
         ​.add(hostnameC, "ALLOW")
         ​.allowUnKnownHostNames(False)
         .build();

```

I agree this is complex. This is because OkHTTP's implementation is not a 'simple CertificatePinner' as you mention. Most pinning implementations just have a list of pins. If the chain is valid and the pin is verified, the connection goes through, while all other connections fail. This is a simple and robust design.
However okhttp has feature that is considered 'advanced' for most pinning implementations out there: hostname-to-pin association. This is a good thing to have; however it also makes it more complex. Essentially, while 'all other connections fail' is now reverted to 'all other connections succeed'.

Anyway, this can all be avoided simply with some more clarity in the documentation; a good step would be to clearly mention that the DEFAULT instance allows all connections and the same is true if no pins are added. 

There are developers that would expect all non-pinned connections to fail; this is not happening.

**Comment by swankjesse** [comment]
> _“Most pinning implementations just have a list of pins.”_

Got some examples? When we added this I didn't think there were very many implementations!

**Comment by ikoz** [comment]
two examples would be moxie's library, https://github.com/moxie0/AndroidPinning/
and sample owasp code https://www.owasp.org/index.php/Certificate_and_Public_Key_Pinning

I have many others in mind but these are in non-public code (i do security code-reviews as a day job and pinning implementations come up a lot, that's how I noticed this okhttp behavior)

I can certainly say that there aren't many pinning implementations that do hostname-to-pin association. 

The main difference I see between most pinning implementations and okhttp is most other implementation have a Pinner that is supposed to be used only for the 'secure' connections.

The concept of using the same handler, that may or may not have a Pinner, for both pinned and non-pinned connections is interesting and is enabled exactly because okhttp has this hostname-to-pin mapping. This is exactly what creates this complications as to what should be the 'default' behavior. 

When I started this thread I thought the default behavior should be to deny all other connections; however I see how this can break some use cases. The next best thing would be to make it configurable, thus the extra proposed options. 

I believe it is a valid use case that someone will want only a few specific hostnames to be pinned and all non-pinned domains to fail. This way he can be certain that no 'non-pinned' connections will be made by his app - which is a critical issue for certain types of apps.
